### ISSUE TITLE
Another pattern for exporting main command

### ISSUE THREAD

**Issue by nxy7** [nushell/nushell#9274]
### Related problem

Hello, some time ago while creating Nushell scripts I thought that Nushell could use another patter for creating scripts that would make it work more like common CLI tools. 

I have the following script file
```nu
#!/usr/bin/env nu

# Utility functions for starting and stopping developer environment
export def main [] {
    help commands main
}

# creates developer environment by running manifests at the root of the project
export def `main start` [] {
    "starting kubernetes"
}

# stops developer environment by deleting resources created by manifests at the root of the project
export def `main stop` [] {
    "stopping kubernetes"
}
```
which produces the following output when run
![image](https://github.com/nushell/nushell/assets/32909532/90829b75-5606-44ea-a367-0706576b711e)
As you can see it's clearly meant to be a place that aggregates other subcommands and there's no real main function and I think help message could be better in that case. Ideally there'd be no 'usage' and I think subcommands could drop 'main' prefix. I'd like my scripts to be easily invoked from shells other than nushell like `nu script_file.nu start` and `main` prefix seems confusing to someone that might not be familliar with nu.

Ideal output:
![image](https://github.com/nushell/nushell/assets/32909532/39b673b2-9cca-4492-a52f-2becb7f84d1c)

I think that  kind of usage pattern would be easier to understand for non nushell users.

### Describe the solution you'd like

I think there are 2 distinct features I'm asking for. 
1) Empty commands that have no 'usage' section and display help message instead
2) Getting rid of 'main' prefix in help messages and in subcommands usage

I think 1) is doable and might be useful for more cases than just mine. Right now im running `help commands main` to simulate that behaviour, but special syntax could signal that this command is not 'usable'. I thought of something like 
```nu
export def main
export def `main echo` [message] {
   $"($message)"
}
export def `main greet`
export def `main greet hi` [name] {
   $"Hi ($name)"
}
export def `main greet hello` [name] {
   $"Hello ($name)"
}
```
So any function with no body would be meant to aggregate subcommands which is common pattern for many CLI tools (and I think nushell would be amazing in doing that kind of work).

Second request would be special treatment for `main` function. I don't really see the best solution, just workflow that in my opinion would be more welcoming. Right now there's different behaviour when invoking nushell script directly and when running `use script.nu` on it. The first script I've posted lives at 'kubernetes.nu` file. If I use it directly `nu kubernetes.nu` it displays help just fine. If I run `use kubernetes.nu` it translates `main` to `kubernetes` in function name so running `kubernetes` doesn't work as `help commands main` (creating empty functions with help messages could solve that issue) points to a function that doesn't exist. That's other problem, but what I'd like is:
a) if function name is `main` drop the prefix from help messages and subcommands
b) use file name instead of `name` if invoked from CLI not with `use script.nu` which doesn't seem that weird considering that nushell already changes function name when running `use`

I think option A would be better and lead to more consistent user experience. Here's usage differs between running `use` and running script directly.
```bash
use kubernetes.nu
nu kubernetes.nu #displays help message for example i've posted
kubernetes #doesn't work as there's no 'main' command anymore after `use`
nu kubernetes.nu start -h #works as intended 
kubernetes start -h #doesn't work as it's main subcommand
kubernetes main start -h #this one works, but doesn't seem consistent with running script file directly
```

Hope my message wasn't very chaotic ^_^ Let me know if it's something You think would be valuable to nushell.

**Comment by kubouch** [comment]
Can you simply define the commands without the `main` prefix, like `export def start [] { ... }` etc.? Then calling `use kubernetes.nu` would give you the correct subcommand structure.

**Comment by nxy7** [comment]
As I've stated in the issue, I want those scripts to be easily sharable with people not using nu, so I'd want to allow access patterns like `nu script.nu command` or `./script.nu command`. Running `use` isn't exactly useful here as people running it will invoke it from shells other than nushell. 
Additionally adding one-off commands to your shell context seems more like a hack than a real solution.

**Comment by kubouch** [comment]
I see, that's an interesting idea. I'm not a big fan of special-casing the `main` prefix because it breaks the mental model around modules and subcommands. But maybe we could have a `run-module` command that would run a module file/directory as a command supporting the subcommands. This could be also exposed via a `--run` flag so you'd be able to do `nu --run-module kubernetes.nu start --something` -- something along these lines.

A current workaround might be to add a dispatch into `main` via a parameter. It's not ideal since you'd lose the subcommand help and passing parameters to the subcommands will be annoying, but maybe could work for simple use cases:
```
export def main [subcommand?: string] {
	match $subcommand {
		"start" => start,
		"stop" => stop,
		_ => help commands main
	}
}
```

You could also call it with `nu -c 'use kubernetes.nu; kubernetes start'`. Again, not ideal, just to give some workarounds.

**Comment by nxy7** [comment]
Yes I'm aware of `nu -c` hack, but I feel like the perfect solution is so close and it's such a good use case for nushell that it's a shame it's not supported natively yet ^_^
Honestly I feel like worse help messages are better than using match statement or `nu -c`.
You're `--run-module` flag solution would be great. I think with a shebang `#!/bin/usr/env nu --run-module` it would allow for super easy direct access while also (possibly) not changing nushell internals and maybe being easier for maintainers to work with :-)

**Comment by tminich** [comment]
After reading the module docs in detail I actually think we're already almost there.

```nushell
#!/usr/bin/env nu

module test {
    export def "main" [] {
        print 'main'
        cat test.nu
    }

    export def "main foo" [] {
        print 'foo'
    }
}

use test *
```

This works for sourcing, but unfortunately not for direct execution.
If this could be made to work 'right' when called via shebang, maybe with a special use syntax, that part of the issue would be solved.

**Comment by nxy7** [comment]
Hmmm, I was thinking about this, is there any reason why 'main' is actually special?
Given the following simple script
```nu
#!/usr/bin/env nu
print "init no main"

export def main [] {
  print "init with main"
}

export def print-hello [name: string] {
  print $name
}

export def add [n1: int, n2: int] {
  print ($n1 + $n2)
}
```
It always invokes the first print "init no main".
Also if main is present it also is always ran (after all code outside of function scopes).
Is that intended behaviour? Because to me if there's `main` function I'd personally ignore top level scripts and only run 'main'.

Back to the topic. What if `main` wasn't special at all and instead all exported commands would be available to use when invoking nushell scripts? Here's how I see it:
1) when running script file check if there are any exported commands
  a) if there are none just run those scripts normally and exit
  b) any exported commands present? Treat as 'file' and ignore top level scripts (or run them first as it does currently if that's intended behaviour, but if I see main i expect it to run first) 
2) now check how user invoked the file.
  a) If he invoked it with subcommand (`./file.nu print-hello David`) that's exported just run it
  b) Was it invoked without subcommand? Run 'main'
  c) 'main' not exported or not present? Show help message.
 
I think that would lead to superior scripting experience, good discoverability and nice syntax. Imagine file like math.nu
```nu
#!/usr/bin/env nu

export def add [n1: int, n2: int] {
   return ($n1 + $n2)
}

export def subtract [n1: int, n2: int] {
   return ($n1 - $n2)
}
```
Now, you don't even need to be familiar with nu, to know what's going on. Syntax is self explanatory - 'export definition add' works as expected (it exports it and shows it to the outside), you lose one 'level' of nesting on subcommands (no need to do this whole 'export def `main add`' business) and I think you don't lose anything. 

In summary. Right now 'main' is treated specially because it's the only thing that gets exported out of script file. What if it was 'special' only as it would be executed by default, but all 'exported' functions were available to the outside?

**Comment by tminich** [comment]
Top level statements are (to me) initializers for the whole script and should be always run. For example if you need to resolve the path to a config file that all (or at least some) subcommands need, that's where it (imo) goes. Of course you cold put that into a utility function to be called by the subcommands that need it, but, at least to me, that seems overly complex.
Another use case could be asserting some basic conditions all subcommands need met to work.

Personally I would not put any top level statements between exported function / main / subcommands, but that is a matter of style.

**Comment by jntrnr** [comment]
We have a possible improvement to the help output here:

https://github.com/nushell/nushell/pull/9948

**Comment by nxy7** [comment]
> We have a possible improvement to the help output here:
> 
> #9948

That's great on the user side, just now I think writing (and reading) scripts could also be simplified. There might be some technical reasons why not all functions are exported from a file, but to me that seems like a clear win. If you want to hide some of them you could put them in another file and not reexport them. 

If #9948 was implemented, then I would not open this issue in the first place, but now I'm curious what's the logic behind scoping functions exported from files to 'main' and it's children.

**Comment by jntrnr** [comment]
@nxy7 - one reason is that it follows how we do subcommands in other places, so once someone learns that, they can reuse it to make script subcommands. Another helpful aspect is that because the help system understands subcommands like this, it can autogenerate the help in the same form users would be used to.

We could possibly also support modules being run directly and treating exports as commands, but I see that as related but different to the main in script thing.

**Comment by texastoland** [comment]
I suspect the root XY problem is there's a mismatch between script subcommands and module subcommands tracked in #11440 (just linking them together).

**Comment by caemanwaves** [comment]
This is the best I've come up with:
```nushell
#!/usr/bin/env nu
module main {
    export def sub [] { 2 }
}
export use main *
use main

# main help
export def main [num=1] { $num }
```
This correctly handles autocomplete, and runs as both script and module but:
* it's messy
* `./cmd.nu -h` and `cmd -h` do not find the sub command

**Comment by caemanwaves** [comment]
Actually, this is better.
```nushell
#!/usr/bin/env nu

# main help
export def main [] { 1 }

# sub help
def 'main sub' [] { 2 }
export alias sub = main sub
```
